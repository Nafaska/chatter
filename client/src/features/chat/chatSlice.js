import { createSlice } from "@reduxjs/toolkit";

export const chatSlice = createSlice({
  name: "chat",
  initialState: {
    message: "",
    channel: [],
    username: ""
  },
  reducers: {
    postMessage: (state, action) => {
      state.message = "";
      state.channel = [
        ...state.channel,
        {
          message: action.payload.data.message,
          time: +new Date(),
          username: action.payload.data.username,
        },
      ];
    },
    typeMessage: (state, action) => {
      state.message = action.payload;
    },
  },
});

export const { postMessage, typeMessage } = chatSlice.actions;

export const selectMessage = (state) => state.chat.message;
export const selectChannel = (state) => state.chat.channel;

export default chatSlice.reducer;
