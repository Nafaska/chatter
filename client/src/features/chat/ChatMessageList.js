import Spinner from "../helpers/Spinner";
import { topography } from "hero-patterns";
import { selectUsername } from "../auth/authSlice";
import { useSelector } from "react-redux";
import { selectName, selectChannelsContent } from "./chatSlice";
import ChatMessage from "./ChatMessage";

const ChatMessageList = () => {
  const name = useSelector(selectName);
  const channelsContent = useSelector(selectChannelsContent);
  const username = useSelector(selectUsername);
  const style = process.env.NODE_ENV !== "test" ? {
    backgroundImage: topography("#F3D5AD", 0.5),
    height: "calc(100% - 9.5rem)",
  } : {
    height: "calc(100% - 9.5rem)",
  };

  return (
    <div
      id="channel"
      data-testid="channel-content"
      style={style}
      className="backgroundStyle border-b box-border px-6 pt-4 sm:pb-8 pb-24 overflow-y-auto items-center"
    >
      {!name && <Spinner height="h-full" />}
      {name && channelsContent[name]
        && Object.values(channelsContent[name]).map((it) => {
            return (
              <ChatMessage
                key={it.time + it.username}
                data={it}
                isOwn={username === it.username}
              />
            );
          })}
      <div className="h-1 auto-overflow-anchoring"></div>
    </div>
  );
};

export default ChatMessageList;
