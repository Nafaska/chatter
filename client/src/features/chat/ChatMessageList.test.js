import { screen, render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import ChatMessageList from "./ChatMessageList";
import TestRenderer from "react-test-renderer";
import { storeMessages, openChat } from "./chatSlice";
import { configureStore as realConfigureStore } from "@reduxjs/toolkit";
import authReducer from "../auth/authSlice";
import chatReducer, { LIMIT_OF_MESSAGES } from "./chatSlice";

let store;

beforeEach(() => {
  store = realConfigureStore({
    reducer: {
      auth: authReducer,
      chat: chatReducer,
    },
  });
});

let messages = [];

for (let i = 0; i < LIMIT_OF_MESSAGES; i++) {
  messages.push({
    data: {
      message: `#${i} message`,
      username: `user${i}`,
      channel: "ChecksLimit",
      time: 1613048699105,
    },
  });
}

describe("dispatching content:", () => {
  test("limit messages per channel", () => {
    render(
      <Provider store={store}>
        <ChatMessageList />
      </Provider>
    );
    store.dispatch(openChat({ name: "ChecksLimit", description: "" }));
    messages.map((message) => store.dispatch(storeMessages(message)));
    expect(screen.getByTestId("channel-content").children).toHaveLength(16);
    store.dispatch(
      storeMessages({
        data: {
          message: "it's a message above the limit",
          username: "Shelby",
          channel: "ChecksLimit",
          time: 1613048699105,
        },
      })
    );
    expect(screen.getByTestId("channel-content").children).toHaveLength(16);
    expect(document.body).toMatchSnapshot();
  });

  test("dispatching a new message to existed one", () => {
    render(
      <Provider store={store}>
        <ChatMessageList />
      </Provider>
    );
    store.dispatch(openChat({ name: "General", description: "" }));
    store.dispatch(
      storeMessages({
        data: {
          message: "1st message",
          username: "Viktoriia Kovalenko",
          channel: "General",
          time: 1613048680105,
        },
      })
    );
    store.dispatch(
      storeMessages({
        data: {
          message: "2nd message",
          username: "Shelby",
          channel: "General",
          time: 1613048699105,
        },
      })
    );
    expect(screen.getByTestId("channel-content").children).toHaveLength(3);
    expect(document.body).toMatchSnapshot();
  });

  test("rendering component without messages", () => {
    render(
      <Provider store={store}>
        <ChatMessageList />
      </Provider>
    );
    store.dispatch(openChat({ name: "General", description: "" }));
    expect(screen.getByTestId("channel-content").textContent).toBe("");
  });
});

describe("rendering content from mocked store:", () => {
  const mockStore = configureStore([]);
  let mockedStore;
  let comp;
  beforeEach(() => {
    mockedStore = mockStore({
      chat: {
        name: "MyChannel",
        channelsContent: {
          MyChannel: {
            0: {
              message: "it is a test",
              time: 1612982540145,
              username: "Viktoriia Kovalenko",
              name: "General",
            },
            1: {
              message: "works fine",
              time: 1612982604222,
              username: "Shelby",
              name: "General",
            },
          },
        },
      },
      auth: {
        username: "Viktoria Kovalenko",
      },
    });

    mockedStore.dispatch = jest.fn();

    comp = TestRenderer.create(
      <Provider store={mockedStore}>
        <ChatMessageList />
      </Provider>
    );
  });
  test("rendering page with 2 messages", () => {
    expect(comp).toMatchSnapshot();
  });
});
