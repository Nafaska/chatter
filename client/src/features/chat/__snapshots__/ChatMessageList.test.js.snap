// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dispatching content: dispatching a new message to existed one 1`] = `
<body>
  <div>
    <div
      class="backgroundStyle border-b box-border px-6 pt-4 sm:pb-8 pb-24 overflow-y-auto items-center"
      data-testid="channel-content"
      id="channel"
      style="height: calc(100% - 9.5rem);"
    >
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              Viktoriia Kovalenko
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:40 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            1st message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              Shelby
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            2nd message
          </p>
        </div>
      </div>
      <div
        class="h-1 auto-overflow-anchoring"
      />
    </div>
  </div>
</body>
`;

exports[`dispatching content: limit messages per channel 1`] = `
<body>
  <div>
    <div
      class="backgroundStyle border-b box-border px-6 pt-4 sm:pb-8 pb-24 overflow-y-auto items-center"
      data-testid="channel-content"
      id="channel"
      style="height: calc(100% - 9.5rem);"
    >
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user1
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #1 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user2
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #2 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user3
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #3 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user4
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #4 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user5
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #5 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user6
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #6 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user7
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #7 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user8
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #8 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user9
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #9 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user10
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #10 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user11
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #11 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user12
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #12 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user13
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #13 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              user14
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            #14 message
          </p>
        </div>
      </div>
      <div
        class="flex items-start mb-4 no-overflow-anchoring"
      >
        <img
          alt="user avatar"
          class="w-12 h-12 rounded-full mr-3"
          src="catAvatar.png"
        />
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-end"
          >
            <span
              class="font-bold text-md mr-2"
              data-testid="username"
            >
              Shelby
            </span>
            <span
              class="text-grey-700 text-xs font-light"
              data-testid="time"
            >
              2/11/2021, 2:04:59 PM
            </span>
          </div>
          <p
            class="font-light break-words text-md text-grey-800 pt-1"
            data-testid="message"
          >
            it's a message above the limit
          </p>
        </div>
      </div>
      <div
        class="h-1 auto-overflow-anchoring"
      />
    </div>
  </div>
</body>
`;

exports[`rendering content from mocked store: rendering page with 2 messages 1`] = `
<div
  className="backgroundStyle border-b box-border px-6 pt-4 sm:pb-8 pb-24 overflow-y-auto items-center"
  data-testid="channel-content"
  id="channel"
  style={
    Object {
      "height": "calc(100% - 9.5rem)",
    }
  }
>
  <div
    className="flex items-start mb-4 no-overflow-anchoring"
  >
    <img
      alt="user avatar"
      className="w-12 h-12 rounded-full mr-3"
      src="catAvatar.png"
    />
    <div
      className="flex flex-col"
    >
      <div
        className="flex items-end"
      >
        <span
          className="font-bold text-md mr-2"
          data-testid="username"
        >
          Viktoriia Kovalenko
        </span>
        <span
          className="text-grey-700 text-xs font-light"
          data-testid="time"
        >
          2/10/2021, 7:42:20 PM
        </span>
      </div>
      <p
        className="font-light break-words text-md text-grey-800 pt-1"
        data-testid="message"
      >
        it is a test
      </p>
    </div>
  </div>
  <div
    className="flex items-start mb-4 no-overflow-anchoring"
  >
    <img
      alt="user avatar"
      className="w-12 h-12 rounded-full mr-3"
      src="catAvatar.png"
    />
    <div
      className="flex flex-col"
    >
      <div
        className="flex items-end"
      >
        <span
          className="font-bold text-md mr-2"
          data-testid="username"
        >
          Shelby
        </span>
        <span
          className="text-grey-700 text-xs font-light"
          data-testid="time"
        >
          2/10/2021, 7:43:24 PM
        </span>
      </div>
      <p
        className="font-light break-words text-md text-grey-800 pt-1"
        data-testid="message"
      >
        works fine
      </p>
    </div>
  </div>
  <div
    className="h-1 auto-overflow-anchoring"
  />
</div>
`;
